experimental/parser/testdata/parser/def/ordering.proto

```diff
@@ -0,0 +1,41 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto2";
+
+package test;
+
+message M {
+    M x (T) (T);
+    M x returns (T) (T);
+    M x returns T (T);
+    M x [foo = bar] (T);
+    M x { /* ... */ } (T);
+
+    M x returns (T) returns (T);
+    M x [foo = bar] returns (T);
+    M x { /* ... */ } returns (T);
```

Comment:
As mentioned before, This looks to me like it would rather be parsed as `M x { }` followed by `returns (T);`. For one, I don't think this particular error is that likely; and two, I think it would make `parseDef` easier to read and understand.

Reply:
This is a much more interesting issue when the thing after the `{}` is a compact options. Making it work just for compact options but not for the others is probably more work than what I've done in the refactor of parseDef.